Cryptolib/ToMathlib.lean

```diff
@@ -0,0 +1,216 @@
+/-
+Copyright (c) 2021 Joey Lupo
+Released under Apache 2.0 license as described in the file LICENSE-APACHE.
+Authors: Joey Lupo, Alfredo Garcia
+-/
+import Mathlib.Data.BitVec
+import Mathlib.Data.Fintype.BigOperators
+import Mathlib.Data.ZMod.Basic
+import Mathlib.GroupTheory.SpecificGroups.Cyclic
+import Mathlib.GroupTheory.Subgroup.Simple
+-- import Mathlib.GroupTheory.Subgroup.Pointwise
+import Mathlib.GroupTheory.OrderOfElement
+import Mathlib.Probability.Distributions.Uniform
+import Mathlib.Probability.ProbabilityMassFunction.Basic
+import Mathlib.Probability.ProbabilityMassFunction.Constructions
+import Mathlib.Probability.ProbabilityMassFunction.Monad
+
+/-!
+# General lemmas for inclusion into mathlib
+
+These are lemmas necessary for using Mathlib with cryptography, that should be upstreamed
+into Mathlib itself.
+-/
+
+/-
+ ---------------------------------------------------------
+  To multiset.range
+ ---------------------------------------------------------
+-/
+
+lemma range_pos_ne_zero (n : ℕ) (n_pos : 0 < n) : Multiset.range n ≠ 0 := by
+  apply (Multiset.card_pos).mp
+  rw [Multiset.card_range]
+  exact n_pos
+
+
+
+/-
+ ---------------------------------------------------------
+  To group_theory.is_cyclic
+ ---------------------------------------------------------
+-/
+
+def is_cyclic.generator {G : Type} [Group G] [IsCyclic G] (g : G): Prop :=
+   ∀ (x : G), x ∈ Subgroup.zpowers g
+
+
+/-
+ ---------------------------------------------------------
+  To bitvec.basic
+ ---------------------------------------------------------
+-/
+
+namespace BitVec
+
+instance fintype : Π (n : ℕ), Fintype (BitVec n) := by
+  intro n
+  sorry
+  -- exact Vector.fintype
+
+lemma card (n : ℕ) : Fintype.card (BitVec n) = 2^n := by sorry
+  -- card_vector n
+
+lemma multiset_ne_zero (n : ℕ) : (BitVec.fintype n).elems.val ≠ 0 := by
+  apply (Multiset.card_pos).mp
+  -- have h : Multiset.card (Fintype.elems (BitVec n)).val = 2^n := BitVec.card n
+  -- rw [h]
+  -- simp only [pow_pos, nat.succ_pos']
+  sorry
+
+-- missing bitvec lemmas used in streams ciphers.
+-- TODO: they need proof
+variable (n : ℕ)
+variable (a b c : BitVec n)
+
+lemma add_self : a + a = BitVec.zero n := by sorry
+-- lemma add_assoc : a + b + c = a + (b + c) := by sorry
+-- lemma zero_add : a = BitVec.zero n + a := by sorry
+lemma add_self_assoc : b = a + (a + b) := by sorry
+  -- rw [←add_assoc, add_self, ←zero_add]
+
+-- lemma add_comm : a + b = b + a := by
+--   -- idea: convert a and b to ℕ and prove comm there
+--   have ha := bitvec.to_nat a
+--   have hb := bitvec.to_nat b
+--   sorry
+
+lemma add_assoc_self : b = a + b + a := by
+  -- rw [add_comm, ←add_assoc, add_self, ←zero_add]
+  sorry
+
+lemma add_assoc_self_reduce : c = a + (b + (a + (b + c))) := by
+  rw [←add_assoc, ←add_assoc, ←add_assoc]
+  -- rw [←add_assoc_self, add_self, ←zero_add]
+  sorry
+
+-- def to_list (length: ℕ) (B : BitVec length) : List Bool :=
+--   Vector.to_list B
```

Comment:
AFAICT these were added to enable representing stream cipher encryption and decryption with `+` instead of XOR. I'm going to remove them.